crypto/src/merkle.rs

```diff
@@ -18,10 +18,16 @@ pub struct MerkleTree<T> {
 }
 
 impl<T: IntoSchema> IntoSchema for MerkleTree<T> {
+    fn type_name() -> String {
+        format!("{}::MerkleTree<{}>", module_path!(), T::type_name())
+    }
     fn schema(map: &mut MetaMap) {
         map.entry(Self::type_name()).or_insert_with(|| {
             // BFS ordered list of leaf nodes
-            Metadata::Vec(HashOf::<T>::type_name())
+            Metadata::Vec(VecMeta {
+                ty: HashOf::<T>::type_name(),
+                sorted: true,
```

Comment:
correct me if I'm wrong here, but it's sorted, isn't it?

Reply:
Though the ordering methods seem to have never been an issue before, there are something odd about
- ```rust
  impl<U> FromIterator<HashOf<U>> for MerkleTree<U>
  ```
- ```rust
  impl<'a, T> Iterator for BreadthFirstIter<'a, T>
  ```

Try this:
```rust
    #[test]
    fn nodes_5() {
        let tree = (0..5)
            .map(|i| Hash::prehashed([i; Hash::LENGTH]).typed())
            .collect::<MerkleTree<()>>();

        for node in tree.iter() {
            dbg!(node);
        }
    }
```

Expected:
```
            #0: iteration order
            ??: first 2 hex of hash
        ____/\____
      #1          #2
      ??          ??
    __/\__      __/\__
  #3      #4  #5
  ??      ??  04
  /\      /\
#6  #7  #8  #9
00  01  02  03
```

Actual:
```
            #0: iteration order
            e7: first 2 hex of hash
        ____/\____
      #8          #1
      93          1a
    __/\__      __/\__
  #9          #5      #2
  04          f4      f6
              /\      /\
            #7  #6  #4  #3
            00  01  02  03
```

Reply:
this doesn't look good

Reply:
Suggestion: DFS
```
            #0: iteration order
            ??: first 2 hex of hash
        ____/\____
      #1          #8
      ??          ??
    __/\__      __/\__
  #2      #5  #9
  ??      ??  04
  /\      /\
#3  #4  #6  #7
00  01  02  03
```
Otherwise `MerkleTree::add` would not work due to disorder of the leaves

Reply:
However, for some reason this passes:
```rust
    #[test]
    fn add_does_save_the_order() {
        let tree0 = (0..5)
            .map(|i| Hash::prehashed([i; Hash::LENGTH]).typed())
            .collect::<MerkleTree<()>>()
            .add(Hash::prehashed([5; Hash::LENGTH]).typed());
        
        let tree1 = (0..=5)
            .map(|i| Hash::prehashed([i; Hash::LENGTH]).typed())
            .collect::<MerkleTree<()>>();

        for (node0, node1) in tree0.iter().zip(tree1.iter()) {
            assert_eq!(node0.hash(), node1.hash())
        }
    }
```

Reply:
we can resolve this?

Reply:
Yeah, anyway it is sorted.
It might depend on how MerkleTree will be used on the client side, but it is acceptable to me

Reply:
it's important that it's sorted to be able to encode it and consequently sign it predictably. What sdk devs do with it I don't care as long as it's sorted in the encoded representation

Reply:
Something we need to be aware of. I'd add a comment here.

Reply:
Created #2128

Reply:
>  It might depend on how MerkleTree will be used on the client side, but it is acceptable to me

We don't care how sdk devs will be using it as long as it produces the same root hash. Therefore, it needs to be encoded consistently so that the signature verification works. That's why it should be sorted in the serialized form